File: src/ts/magma/app/compile/value/Symbol.ts
import { Type } from "../../../../magma/app/compile/type/Type";
import { CompileState } from "../../../../magma/app/compile/CompileState";
import { Definition } from "../../../../magma/app/compile/define/Definition";
import { PrimitiveType } from "../../../../magma/app/compile/type/PrimitiveType";
import { Node } from "../../../../magma/app/compile/node/Node";
import { Option } from "../../../../magma/api/option/Option";
import { Some } from "../../../../magma/api/option/Some";
import { ValueCompiler } from "../../../../magma/app/ValueCompiler";
export class Symbol {
	value: string;
	constructor (value: string) {
		this.value = value;
	}
	generate(): string {
		return this.value/*unknown*/;
	}
	resolve(state: CompileState): Type {
		return state.stack().resolveNode(this.value).map((definition: Definition) => {
			return definition.findType()/*unknown*/;
		}).orElse(PrimitiveType.Unknown)/*unknown*/;
	}
	toNode(): Option<Node> {
		return new Some<Node>(this)/*unknown*/;
	}
	isFunctional(): boolean {
		return false/*unknown*/;
	}
	isVar(): boolean {
		return false/*unknown*/;
	}
	generateBeforeName(): string {
		return ""/*unknown*/;
	}
	generateSimple(): string {
		return ValueCompiler.generateValue(this)/*unknown*/;
	}
}

File: src/java/magma/app/compile/value/Symbol.java
package magma.app.compile.value;

import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.app.compile.CompileState;
import magma.app.compile.define.Definition;
import magma.app.compile.type.PrimitiveType;
import magma.app.compile.type.Type;

public record Symbol(String value) implements Value, Type {
    @Override
    public String generate() {
        return this.value;
    }

    public Type resolve(CompileState state) {
        return state.stack().resolveValue(this.value)
                .map((Definition definition) -> {
                    return definition.findType();
                })
                .orElse(PrimitiveType.Unknown);
    }

    @Override
    public Option<Value> toValue() {
        return new Some<Value>(this);
    }

    @Override
    public Option<Value> findChild() {
        return new None<Value>();
    }

    @Override
    public boolean isFunctional() {
        return false;
    }

    @Override
    public boolean isVar() {
        return false;
    }

    @Override
    public String generateBeforeName() {
        return "";
    }

    @Override
    public Option<String> generateAsEnumValue(String structureName) {
        return new None<String>();
    }

    @Override
    public String generateSimple() {
        return this.generate();
    }
}
